File: src/mvmctl/core/image/_resolver.py
"""Image resolution helpers."""

from __future__ import annotations

from dataclasses import dataclass

from mvmctl.core._shared import RelationEnricher, RelationSpec
from mvmctl.core.image._repository import ImageRepository
from mvmctl.exceptions import ImageNotFoundError
from mvmctl.models import ImageItem

__all__ = [
    "ImageResolver",
    "ImageResolveResult",
]


@dataclass
class ImageResolveResult:
    items: list[ImageItem]
    errors: list[str]
    exit_code: int


class ImageResolver:
    """Resolver for image resources."""

    RELATIONS: dict[str, RelationSpec] = {
        "vm": RelationSpec(
            fk_field="id",
            resolver="vm",
            method="by_image_id",
            relation_name="vms",
            is_reverse=True,
            batch_method="by_image_id_batch",
        ),
    }

    def __init__(
        self,
        repo: ImageRepository | None = None,
        *,
        include: list[str] | None = None,
    ) -> None:
        self._repo = repo if repo is not None else ImageRepository()
        self._include = include

    def _enrich(self, images: list[ImageItem]) -> list[ImageItem]:
        """Enrich images with relations if include is set."""
        if self._include and images:
            RelationEnricher().enrich(images, self._include, self.RELATIONS)
        return images

    def by_id(self, image_id: str) -> ImageItem:
        """Resolve by full ID."""
        matches = self._repo.find_by_prefix(image_id)
        if len(matches) == 0:
            raise ImageNotFoundError(f"Image not found: {image_id!r}")
        if len(matches) > 1:
            raise ImageNotFoundError(f"Image ID is ambiguous: {image_id!r}")
        return self._enrich(matches)[0]

    def by_os_slug(self, os_slug: str) -> ImageItem:
        """Resolve by OS slug."""
        db_image = self._repo.get_by_os_slug(os_slug)
        if db_image is None:
            raise ImageNotFoundError(f"Image not found: {os_slug!r}")
        return self._enrich([db_image])[0]

    def get_default(self) -> ImageItem | None:
        """Resolve the default image, or None if not set."""
        image = self._repo.get_default()
        if image is None:
            return None
        return self._enrich([image])[0]

    def resolve(self, value: str) -> ImageItem:
        """Resolve image by os_slug or ID prefix."""
        try:
            image = self.by_os_slug(value)
        except ImageNotFoundError:
            image = self.by_id(value)
        return image

    def resolve_many(self, identifiers: list[str]) -> ImageResolveResult:
        """Resolve multiple image identifiers by os_slug or id."""
        # Deduplicate identifiers while preserving order
        seen_inputs: set[str] = set()
        unique_ids: list[str] = []
        for ident in identifiers:
            if ident not in seen_inputs:
                seen_inputs.add(ident)
                unique_ids.append(ident)

        items: list[ImageItem] = []
        errors: list[str] = []
        resolved_ids: set[str] = set()

        for identifier in unique_ids:
            try:
                item = self.resolve(identifier)
                if item.id not in resolved_ids:
                    resolved_ids.add(item.id)
                    items.append(item)
            except Exception as e:
                errors.append(f"{identifier}: {e}")

        items = self._enrich(items)

        exit_code = 1 if errors and not items else 0
        return ImageResolveResult(
            items=items, errors=errors, exit_code=exit_code
        )


from mvmctl.core._shared import register  # noqa: E402

register("image", lambda: ImageResolver)

File: internal/core/image/resolver.go
package image

import (
	"context"
	"fmt"

	"mvmctl/internal/infra"
	"mvmctl/internal/lib/model"
	"mvmctl/internal/lib/version"
	"mvmctl/pkg/errs"
)

// RELATIONS defines the cross-domain relations for image enrichment.
var RELATIONS = map[string]model.RelationSpec{
	"vm": {
		FKField:      "id",
		Resolver:     "vm",
		Method:       "by_image_id",
		RelationName: "vms",
		IsReverse:    true,
		BatchMethod:  "by_image_id_batch",
	},
}

// ResolveResult holds the result of a batch image resolution.
type ResolveResult struct {
	Items    []*model.ImageItem
	Errors   []string
	ExitCode int
}

// EnrichFunc is a function that enriches images in-place with relations.
// Set by the API layer during wiring to avoid circular imports.
type EnrichFunc func(ctx context.Context, images []*model.ImageItem, include []string, relations map[string]model.RelationSpec)

// Resolver resolves images by ID, type:version, type, name, or ID prefix.
type Resolver struct {
	repo       Repository
	include    []string
	enrichFunc EnrichFunc
}

// NewResolver creates a new Resolver.
func NewResolver(repo Repository) *Resolver {
	return &Resolver{repo: repo}
}

// SetEnrichFunc sets the enrichment function called after each resolution.
// Must be set for cross-domain relation enrichment (e.g., populating VMs).
func (r *Resolver) SetEnrichFunc(fn EnrichFunc) {
	r.enrichFunc = fn
}

// enrich enriches images with relations if include is set.
func (r *Resolver) enrich(ctx context.Context, images []*model.ImageItem) []*model.ImageItem {
	if r.include != nil && len(images) > 0 && r.enrichFunc != nil {
		r.enrichFunc(ctx, images, r.include, RELATIONS)
	}
	return images
}

// SetInclude sets the relation names to include during resolution.
func (r *Resolver) SetInclude(include []string) {
	r.include = include
}

// ByID resolves by full ID.
func (r *Resolver) ByID(ctx context.Context, imageID string) (*model.ImageItem, error) {
	matches, err := r.repo.FindByPrefix(ctx, imageID)
	if err != nil {
		return nil, fmt.Errorf("resolve image by ID: %w", err)
	}
	if len(matches) == 0 {
		return nil, errs.NotFound(errs.CodeImageNotFound, fmt.Sprintf("Image not found: '%s'", imageID))
	}
	if len(matches) > 1 {
		return nil, errs.NotFound(errs.CodeImageNotFound, fmt.Sprintf("Image ID is ambiguous: '%s'", imageID))
	}
	return r.enrich(ctx, matches)[0], nil
}

// ByVersionType resolves by version and type (both required).
func (r *Resolver) ByVersionType(ctx context.Context, version, imgType string) (*model.ImageItem, error) {
	dbImage, err := r.repo.GetByVersionAndType(ctx, version, imgType)
	if err != nil {
		return nil, err
	}
	if dbImage == nil {
		return nil, errs.NotFound(
			errs.CodeImageNotFound,
			fmt.Sprintf("Image not found: version='%s', type='%s'", version, imgType),
		)
	}
	return r.enrich(ctx, []*model.ImageItem{dbImage})[0], nil
}

// ByType resolves by image type.
func (r *Resolver) ByType(ctx context.Context, imgType string) (*model.ImageItem, error) {
	dbImage, err := r.repo.GetByType(ctx, imgType)
	if err != nil {
		return nil, err
	}
	if dbImage == nil {
		return nil, errs.NotFound(errs.CodeImageNotFound, fmt.Sprintf("Image not found: '%s'", imgType))
	}
	return r.enrich(ctx, []*model.ImageItem{dbImage})[0], nil
}

// GetDefault resolves the default image, or nil if not set.
func (r *Resolver) GetDefault(ctx context.Context) (*model.ImageItem, error) {
	image, err := r.repo.GetDefault(ctx)
	if err != nil {
		return nil, err
	}
	if image == nil {
		return nil, nil
	}
	return r.enrich(ctx, []*model.ImageItem{image})[0], nil
}

// ByName resolves by display name.
func (r *Resolver) ByName(ctx context.Context, name string) (*model.ImageItem, error) {
	dbImage, err := r.repo.GetByName(ctx, name)
	if err != nil {
		return nil, err
	}
	if dbImage == nil {
		return nil, errs.NotFound(errs.CodeImageNotFound, fmt.Sprintf("Image not found by name: '%s'", name))
	}
	return r.enrich(ctx, []*model.ImageItem{dbImage})[0], nil
}

// isImageNotFoundError checks if the error is an ImageNotFoundError
// using direct type assertion — no unwrapping.
func isImageNotFoundError(err error) bool {
	de, ok := err.(*errs.DomainError)
	if !ok {
		return false
	}
	return de.Code == errs.CodeImageNotFound
}

// Resolve resolves image by type:version, type, display name, or ID prefix.
// Only ImageNotFoundError causes fallthrough to the next resolution method
// — all other errors propagate immediately.
func (r *Resolver) Resolve(ctx context.Context, value string) (*model.ImageItem, error) {
	// Try "type:version" selector format first using the shared version parser.
	name, ver := version.ParseSelector(value)
	if name != "" && ver != "" {
		image, err := r.ByVersionType(ctx, ver, name)
		if err == nil {
			return image, nil
		}
		if !isImageNotFoundError(err) {
			return nil, err
		}
		// Fall through to type-only lookup with the type part
		value = name
	}

	image, err := r.ByType(ctx, value)
	if err == nil {
		return image, nil
	}
	if !isImageNotFoundError(err) {
		return nil, err
	}

	image, err = r.ByName(ctx, value)
	if err == nil {
		return image, nil
	}
	if !isImageNotFoundError(err) {
		return nil, err
	}

	return r.ByID(ctx, value)
}

// ResolveMany resolves multiple image identifiers.
func (r *Resolver) ResolveMany(ctx context.Context, identifiers []string) *ResolveResult {
	// Dedup input identifiers (e.g. duplicate CLI args) before processing.
	uniqueIDs := infra.Dedup(identifiers)

	var items []*model.ImageItem
	var errorsList []string
	resolvedIDs := make(map[string]bool)

	for _, identifier := range uniqueIDs {
		item, err := r.Resolve(ctx, identifier)
		if err != nil {
			errorsList = append(errorsList, err.Error())
			continue
		}
		if !resolvedIDs[item.ID] {
			// Dedup by DB record ID — different input identifiers may
			// resolve to the same DB record (e.g. "image" and
			// "image:ubuntu" when only ubuntu exists).
			resolvedIDs[item.ID] = true
			items = append(items, item)
		}
	}

	items = r.enrich(ctx, items)

	exitCode := 0
	if len(errorsList) > 0 && len(items) == 0 {
		exitCode = 1
	}

	return &ResolveResult{
		Items:    items,
		Errors:   errorsList,
		ExitCode: exitCode,
	}
}
